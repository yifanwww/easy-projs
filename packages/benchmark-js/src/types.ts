/**
 * The type representing millisecond.
 */
export type Millisecond = number & {};

/**
 * The type representing nanosecond.
 *
 * The maximum safe integer in Javascript is `2^53 - 1` (`Number.MAX_SAFE_INTEGER`), can represent nanosecond.
 */
export type Nanosecond = number & {};

export type TestFn = (...args: never[]) => void;

export interface BenchmarkJobCallbacks {
    /**
     * Called when benchmark starts running.
     */
    onStart?: () => void;
    /**
     * Called when the benchmark completes running.
     */
    onComplete?: () => void;
}

export interface BenchmarkJobSettings {
    /**
     * The delay between test cycles (ms).
     *
     * Default is `5`.
     */
    delay?: Millisecond;
    /**
     * The initial number of ops to run in a benchmark.
     *
     * Default is `16`.
     */
    initOps?: number;
    /**
     * The minimum time a benchmark uses.
     *
     * Default is `100`.
     */
    minSampleTime?: Millisecond;
    /**
     * The count of samples required to perform statistical analysis.
     *
     * Default is `15`.
     */
    samplesCount?: number;
}

/**
 * An array of the choices a argument can use.
 */
export type TestFnArgumentValues = unknown[];
/**
 * An array contains several arrays of the choices a argument can use.
 *
 * Example:
 * ```ts
 * [[1, 2], [true], ['a', 'b']]
 * ```
 */
export type TestFnArgumentsValues = Array<TestFnArgumentValues | undefined>;

export interface BenchmarkJobTestFnOptions {
    /**
     * Used for adjust-benchmarking and fomal-benchmarking.
     *
     * Example:
     * ```ts
     * [[1, 2], [true], ['a', 'b']]
     * ```
     */
    args?: TestFnArgumentsValues;
    /**
     * Used for pre-benchmarking. The arguments provided in `args` will be also added into `preArgs`.
     *
     *
     * Example:
     * ```ts
     * [[1, 2], [true], ['a', 'b']]
     * ```
     */
    preArgs?: TestFnArgumentsValues;
}

export interface BenchmarkJobOptions extends BenchmarkJobCallbacks, BenchmarkJobSettings, BenchmarkJobTestFnOptions {}
