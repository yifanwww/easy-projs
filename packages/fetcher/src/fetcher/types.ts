import type { QueryURLParamsInit } from './url.js';

type Method =
    | 'get'
    | 'post'
    | 'put'
    | 'patch'
    | 'options'
    | 'delete'
    | 'head'
    | 'GET'
    | 'POST'
    | 'PUT'
    | 'PATCH'
    | 'OPTIONS'
    | 'DELETE'
    | 'HEAD';

type FetchResponseType = 'arrayBuffer' | 'blob' | 'formData' | 'json' | 'stream' | 'text';

export interface FetchFactoryOptions {
    /**
     * Authorization header value to send with the request.
     */
    auth?: string;
    /**
     * A base URL from which to resolve all URLs.
     */
    baseURL?: string;
    /**
     * Request headers.
     */
    headers?: Record<string, string>;
    /**
     * Customized function to stringify querystring parameters.
     */
    paramsSerializer?: (params: QueryURLParamsInit) => string;
    /**
     * Customized function to handle status code. By default 200-399 is success.
     */
    validateStatus?: (status: number) => boolean;
    /**
     * Send the request with credentials like cookies.
     */
    withCredentials?: boolean;
}

export interface FetchOptions<T extends BodyInit | object = object> {
    /**
     * Authorization header value to send with the request.
     */
    auth?: string;
    /**
     * Request headers.
     */
    headers?: Record<string, string>;
    data?: T;
    method?: Method;
    /**
     * querystring parameters.
     */
    params?: QueryURLParamsInit;
    /**
     * An encoding to use for the response. Default is `json`.
     */
    responseType?: FetchResponseType;
    /**
     * Customized function to stringify querystring parameters.
     */
    paramsSerializer?: (params: QueryURLParamsInit) => string;
    /**
     * Abort signal.
     */
    signal?: AbortSignal;
    /**
     * Customized function to handle status code. By default 200-399 is success.
     */
    validateStatus?: (status: number) => boolean;
    /**
     * Send the request with credentials like cookies.
     */
    withCredentials?: boolean;
}

export interface FetchResponse<Data = unknown> {
    data: Data;
    headers: Headers;
    redirected: boolean;
    status: number;
    statusText: string;
    type: ResponseType;
    url: string;
}
