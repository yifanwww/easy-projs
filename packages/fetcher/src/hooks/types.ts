import type { FetchOptions } from '../fetcher/index.js';

export type HookFetcher<Data> = (options: FetchOptions) => Promise<Data>;

export interface GetHookFetcher<Data> extends HookFetcher<Data> {
    // eslint-disable-next-line @typescript-eslint/naming-convention
    URL: string;
}
