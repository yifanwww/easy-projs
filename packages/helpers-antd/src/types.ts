import type { SelectProps } from 'antd';

export type FieldNames = NonNullable<SelectProps['fieldNames']>;
