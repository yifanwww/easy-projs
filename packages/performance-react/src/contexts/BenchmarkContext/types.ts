import { BenchmarkResult, BenchmarkResultState, BenchmarkTypes, ComponentName } from 'src/common/benchmark';

export interface IBenchmarkContext {
    mount: Record<ComponentName, BenchmarkResultState>;
    unmount: Record<ComponentName, BenchmarkResultState>;
    update: Record<ComponentName, BenchmarkResultState>;
    totalResults: BenchmarkResultState;
}

export interface IBenchmarkContextUpdaters {
    add: (record: BenchmarkResult) => void;
    clear: (benchmarkType: BenchmarkTypes, componentName: ComponentName) => void;
    clearAll: () => void;
}
