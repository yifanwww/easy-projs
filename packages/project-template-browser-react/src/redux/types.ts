import { ReduxReducer } from '@easy/utils-redux';

interface DisplayConfig {
    clientAreaSize: ClientAreaSize;
}

export interface StoreState {
    displayConfig: DisplayConfig;
    prepared: boolean;
}

export type MainReducer<Payload = undefined> = ReduxReducer<StoreState, Payload>;
