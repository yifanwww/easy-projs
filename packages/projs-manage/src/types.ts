export interface Author {
    email?: string;
    name?: string;
    url?: string;
}

export interface Bugs {
    email?: string;
    url?: string;
}

export interface Repository {
    diectory?: string;
    type?: string;
    url?: string;
}

export interface Engines {
    node?: string;
    vscode?: string;
}

export interface PackageInfo {
    name: string;
    version: string;
    description: string;
    main?: string;
    module?: string;
    types?: string;
    typings?: string;
    homepage?: string;
    license?: string;
    author?: string | Author;
    bugs?: Bugs;
    private?: boolean;
    repository?: Repository;
    scripts: Record<string, string>;
    engines?: Engines;
    dependencies?: Record<string, string>;
    devDependencies?: Record<string, string>;
    workspaces?: string[];
}
