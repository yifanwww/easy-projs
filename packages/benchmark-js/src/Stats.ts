import { tTable } from './constants';
import { Formatter } from './Formatter';
import { getMean, getVariance } from './tools';
import { _Nanosecond } from './types.internal';

/**
 * Class for stats including mean, margin or error, and standard deviation.
 */
export class Stats {
    /**
     * The array of sampled periods.
     */
    private _samples: _Nanosecond[];
    /**
     * The sample arithmetic mean (secs).
     */
    private _mean: _Nanosecond;

    /**
     * The sample standard deviation.
     */
    private _deviation: number;
    /**
     * The margin of error.
     */
    private _moe: number;
    /**
     * The relative margin of error (expressed as a percentage of the mean).
     */
    private _rme: number;
    /**
     * The standard error of the mean (a.k.a. the standard deviation of the sampling distribution of the sample mean).
     */
    private _sem: number;
    /**
     * The sample variance.
     */
    private _variance: number;

    /**
     * The number of executions per second.
     */
    private _ops: number;

    public constructor(samples: _Nanosecond[]) {
        this._samples = samples;

        const size = samples.length;
        this._mean = getMean(samples);
        this._variance = getVariance(samples, this._mean);
        this._deviation = Math.sqrt(this._variance);
        this._sem = this._deviation / Math.sqrt(size);
        const critical = tTable[size - 1] ?? tTable.infinity;
        this._moe = this._sem * critical;
        this._rme = (this._moe / this._mean) * 100 ?? 0;

        this._ops = 1e9 / this._mean;
    }

    public get samples(): _Nanosecond[] {
        return this._samples;
    }

    public get rme(): number {
        return this._rme;
    }

    public get ops(): number {
        return this._ops;
    }

    public toString(order?: number): string {
        const size = this._samples.length;

        const opsStr = Formatter.beautifyNumber(this._ops.toFixed(this._ops < 100 ? 2 : 0));
        const rmeStr = this._rme.toFixed(2);

        return [
            order ? `${order} >` : '',
            `${opsStr} ops/sec`,
            ` ${rmeStr}%`,
            ` (${size} sample${size > 1 ? 's' : ''})`,
        ].join('');
    }
}
