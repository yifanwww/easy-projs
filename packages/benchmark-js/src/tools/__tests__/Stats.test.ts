import { expectSnapshot } from '@easy/utils-test';

import { _Nanosecond } from '../../types.internal';
import { Stats } from '../Stats';

describe(`test class \`${Stats.name}\``, () => {
    it('calculates', () => {
        const stats = new Stats('name', [
            1, 2, 3, 4, 3, 23, 21, 2, 3, 3, 2, 21, 1, 2, 23, 3, 4, 4, 32,
        ] as _Nanosecond[]);

        expectSnapshot(JSON.stringify(stats, undefined, 4));

        expectSnapshot(stats.toString());
        expectSnapshot(stats.toString(2));
    });
});
