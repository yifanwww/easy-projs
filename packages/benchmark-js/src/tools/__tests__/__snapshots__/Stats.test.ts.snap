// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test class \`Stats\` calculates 1`] = `121019108.28025477`;

exports[`test class \`Stats\` calculates 2`] = `57.96879268569932`;

exports[`test class \`Stats\` calculates 3`] = `
Array [
  1,
  2,
  3,
  4,
  3,
  23,
  21,
  2,
  3,
  3,
  2,
  21,
  1,
  2,
  23,
  3,
  4,
  4,
  32,
]
`;

exports[`test class \`Stats\` calculates 4`] = `
"{
    \\"_samples\\": [
        1,
        2,
        3,
        4,
        3,
        23,
        21,
        2,
        3,
        3,
        2,
        21,
        1,
        2,
        23,
        3,
        4,
        4,
        32
    ],
    \\"_mean\\": 8.263157894736842,
    \\"_deviation\\": 9.93781836813939,
    \\"_moe\\": 4.790052869291997,
    \\"_rme\\": 57.96879268569932,
    \\"_sem\\": 2.2798918939990465,
    \\"_variance\\": 98.76023391812865,
    \\"_ops\\": 121019108.28025477
}"
`;

exports[`test class \`Stats\` calculates 5`] = `"121,019,108 ops/sec 57.97% (19 samples)"`;

exports[`test class \`Stats\` calculates 6`] = `"2> 121,019,108 ops/sec 57.97% (19 samples)"`;
